File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> sm;
        unordered_map<char,int> tm;
        for(char c : t){
            tm[c]++;
        }
        int match = tm.size();
        int minLen = s.size()+1;
        int left = 0;
        int right = 0;
        int start =-1;
        while(right<s.size())
        {
            char ch = s[right];
            sm[ch]++;
            if(sm[ch]==tm[ch]) match--;
            while(match==0 && left<=right)
            {
                if(minLen>right+1-left) {
                    start = left;
                    minLen=right+1-left;
                }
                char temp = s[left];
                sm[temp]--;
                left++;
                if(tm.find(temp) != tm.end() && sm[temp]<tm[temp]) match++;
            }
            right++;
        }

        if(start!=-1) return s.substr(start,minLen);
        else return "";
        
    }
};

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";

        Map<Character, Integer> f1 = new HashMap<>();  // Frequency map for t
        Map<Character, Integer> f2 = new HashMap<>();  // Frequency map for s window

        for (char c : t.toCharArray()) {
            f1.put(c, f1.getOrDefault(c, 0) + 1);
        }

        int match = f1.size();
        int left = 0, right = 0;
        int start = 0, len = Integer.MAX_VALUE;

        while (right < s.length()) {
            char c = s.charAt(right);
            f2.put(c, f2.getOrDefault(c, 0) + 1);

            if (f1.containsKey(c) && f1.get(c).intValue() == f2.get(c).intValue()) {
                match--;
            }

            while (match == 0) {
                if (right - left + 1 < len) {
                    start = left;
                    len = right - left + 1;
                }

                char leftChar = s.charAt(left);
                f2.put(leftChar, f2.get(leftChar) - 1);

                if (f1.containsKey(leftChar) && f2.get(leftChar) < f1.get(leftChar)) {
                    match++;
                }

                left++;
            }

            right++;
        }

        return len == Integer.MAX_VALUE ? "" : s.substring(start, start + len);

    }
}